statesync/reactor.go

```diff
@@ -27,189 +62,299 @@ const (
+							Chunks:   snapshot.Chunks,
+							Hash:     snapshot.Hash,
+							Metadata: snapshot.Metadata,
+						},
+					}
+				}
+
+			case *ssproto.SnapshotsResponse:
+				r.mtx.RLock()
+
+				if r.syncer == nil {
+					r.mtx.RUnlock()
+					r.Logger.Debug("received unexpected snapshot; no state sync in progress")
+					continue
+				}
+
+				r.mtx.RUnlock()
+
+				r.Logger.Debug(
+					"received snapshot",
+					"height", msg.Height,
+					"format", msg.Format,
+					"peer", envelope.From.String(),
+				)
+				_, err := r.syncer.AddSnapshot(envelope.From, &snapshot{
+					Height:   msg.Height,
+					Format:   msg.Format,
+					Chunks:   msg.Chunks,
+					Hash:     msg.Hash,
+					Metadata: msg.Metadata,
+				})
+				if err != nil {
+					r.Logger.Error(
+						"failed to add snapshot",
+						"height", msg.Height,
+						"format", msg.Format,
+						"err", err,
+						"channel", r.snapshotCh.ID,
+					)
+					continue
+				}
+
+			default:
+				r.Logger.Error("received unknown message: %T", msg)
+				r.snapshotCh.Error <- p2p.PeerError{
+					PeerID:   envelope.From,
+					Err:      fmt.Errorf("received unknown message: %T", msg),
+					Severity: p2p.PeerErrorSeverityLow,
+				}
+			}
+
+		case <-r.snapshotChDone:
+			r.Logger.Debug("stopped listening on snapshot channel")
+			return
+		}
 	}
 }
 
-// Receive implements p2p.Reactor.
-// XXX: do not call any methods that can block or incur heavy processing.
-// https://github.com/tendermint/tendermint/issues/2888
-func (r *Reactor) Receive(chID byte, src p2p.Peer, msgBytes []byte) {
-	if !r.IsRunning() {
-		return
-	}
+func (r *Reactor) processChunkCh() {
+	for {
+		select {
+		case envelope := <-r.chunkCh.In:
+			switch msg := envelope.Message.(type) {
+			case *ssproto.ChunkRequest:
+				r.Logger.Debug(
+					"received chunk request",
+					"height", msg.Height,
+					"format", msg.Format,
+					"chunk", msg.Index,
+					"peer", envelope.From.String(),
+				)
+				resp, err := r.conn.LoadSnapshotChunkSync(context.Background(), abci.RequestLoadSnapshotChunk{
+					Height: msg.Height,
+					Format: msg.Format,
+					Chunk:  msg.Index,
+				})
+				if err != nil {
+					r.Logger.Error(
+						"failed to load chunk",
+						"height", msg.Height,
+						"format", msg.Format,
+						"chunk", msg.Index,
+						"err", err,
+						"peer", envelope.From.String(),
+					)
+					continue
+				}
 
-	msg, err := decodeMsg(msgBytes)
-	if err != nil {
-		r.Logger.Error("Error decoding message", "src", src, "chId", chID, "msg", msg, "err", err, "bytes", msgBytes)
-		r.Switch.StopPeerForError(src, err)
-		return
-	}
-	err = validateMsg(msg)
-	if err != nil {
-		r.Logger.Error("Invalid message", "peer", src, "msg", msg, "err", err)
-		r.Switch.StopPeerForError(src, err)
-		return
-	}
+				r.Logger.Debug(
+					"sending chunk",
+					"height", msg.Height,
+					"format", msg.Format,
+					"chunk", msg.Index,
+					"peer", envelope.From.String(),
+				)
+				r.chunkCh.Out <- p2p.Envelope{
```

Comment:
can this block forever? shouldn't we select from either ^ or `<-chunkChDone`?

Reply:
The shim listens for outbound messages on the p2p `Channel` and calls either `Broadcast` or `Send` on the (legacy) `Peer`.

@erikgrinaker seems to agree here. Note, if we opt for a `select`, then there is a possibility that messages will be dropped. This is likely probable if there is high throughput on this channel.

What if instead, these were buffered? Currently, they are not.

Reply:
This is related to termination mechanisms, see my other comments. Let's discuss this later today.

Reply:
Hmmm I don't think so. The shim listens for outbound messages. If we use `select` and the shim's listener is busy, it'll be dropped.

Reply:
Yeah, but it's very unergonomic if every single send in every reactor has to check for closure. The channel/router infrastructure should make this convenient somehow.